cpp/dolfin/fem/dofs_permutation.cpp

```diff
@@ -429,233 +482,110 @@ hexahedron_rotations_and_reflection(const int volume_dofs, const int blocksize)
+    // entity of dimension dim
+    for (int i = 0; i < mesh::cell_num_entities(type, d); ++i)
+      entities.push_back({f, vertices.row(i)});
+  }
+  { // scope
+    // Add on the cell itself as an entity
+    auto f = get_ordering_function(type);
+    Eigen::Array<int, 1, Eigen::Dynamic> row(num_vertices_per_cell);
+    for (int i = 0; i < num_vertices_per_cell; ++i)
+      row[i] = i;
+    entities.push_back({f, row});
   }
 
-  return cell_orders;
-}
-//-----------------------------------------------------------------------------
-Eigen::Array<std::int8_t, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>
-compute_ordering_interval(const mesh::Mesh& mesh)
-{
-  const int num_cells = mesh.num_entities(mesh.topology().dim());
-  const int num_permutations = get_num_permutations(mesh.cell_type());
+  // Set orders for each cell
   Eigen::Array<std::int8_t, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>
       cell_orders(num_cells, num_permutations);
-
-  // Set orders for each cell
   const std::vector<std::int64_t>& global_indices
       = mesh.topology().global_indices(0);
+  // Reserve memory to store global numbers of vertices of cell
+  std::vector<std::int64_t> cell_vs(num_vertices_per_cell);
   for (int cell_n = 0; cell_n < num_cells; ++cell_n)
   {
-    const mesh::MeshEntity cell(mesh, 1, cell_n);
-    const int* vertices = cell.entities(0);
-    const std::int64_t v0 = global_indices[vertices[0]];
-    const std::int64_t v1 = global_indices[vertices[1]];
+    // Get the cell and info about it
+    const mesh::MeshEntity cell(mesh, tdim, cell_n);
+    const std::int32_t* local_vertices = cell.entities(0);
+    for (int i = 0; i < num_vertices_per_cell; ++i)
+      cell_vs[i] = global_indices[local_vertices[i]];
 
-    // Set the orders for the edge flip
-    cell_orders(cell_n, 0) = (v0 > v1);
-  }
-
-  return cell_orders;
-}
-//-----------------------------------------------------------------------------
-Eigen::Array<std::int8_t, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>
-compute_ordering_quadrilateral(const mesh::Mesh& mesh)
-{
-  const int num_cells = mesh.num_entities(mesh.topology().dim());
-  const int num_permutations = get_num_permutations(mesh.cell_type());
-  Eigen::Array<std::int8_t, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>
-      cell_orders(num_cells, num_permutations);
+    int j = 0;
+    // iterate over the cell's entities
+    for (std::size_t e_n = 0; e_n < entities.size(); ++e_n)
+    {
+      // Use the functions for each entities and that entity's global
+      // vertex numbers to calculate the number of times each
+      // permutation should be applied on this cell
+      auto f = entities[e_n].first;
+      Eigen::Array<int, 1, Eigen::Dynamic> v = entities[e_n].second;
+      Eigen::Array<std::int64_t, 1, Eigen::Dynamic> global_v(v.size());
+      for (int i = 0; i < v.size(); ++i)
+        global_v[i] = cell_vs[v[i]];
+      Eigen::Array<std::int8_t, 1, Eigen::Dynamic> orders = f(global_v);
```

Comment:
This is a very likely performance problem because of the dynamic memory allocation to create `orders`. Change `f` to take `orders` as an in/out argument.